File: .github/atoma/tools/scripts/inject_sub_results.py
#!/usr/bin/env python3
"""
inject_sub_results.py — Replace the last tool message in a session JSON with
aggregated sub-issue completion results.

Usage:
  python3 inject_sub_results.py \\
    --session session.json \\
    --repo OWNER/REPO \\
    --parent N \\
    --sub-issues 2,3,4 \\
    --out session.json

Reads the session, finds the last ``role: "tool"`` message (the one left by
``launch_sub_agent``), and replaces its content with an aggregated result
summarizing the completed sub-issues and their linked PRs.
"""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from typing import Any


def find_last_tool_index(messages: list[dict[str, Any]]) -> int | None:
    """Return the index of the last message with role='tool', or None."""
    for i in range(len(messages) - 1, -1, -1):
        if messages[i].get("role") == "tool":
            return i
    return None


def gather_sub_results(repo: str, sub_issues: list[int]) -> str:
    """Query GitHub for each sub-issue's linked PRs and build a summary."""
    lines: list[str] = []
    lines.append("All sub-issues have been completed.")
    lines.append("")
    lines.append("## Sub-issue Results")
    lines.append("")

    for num in sub_issues:
        try:
            # Get issue title and linked PRs
            result = subprocess.run(
                [
                    "gh", "issue", "view", str(num),
                    "--repo", repo,
                    "--json", "title,state,closedAt",
                ],
                capture_output=True, text=True, timeout=15,
            )
            issue_info = json.loads(result.stdout) if result.returncode == 0 else {}
            title = issue_info.get("title", "Unknown")
            state = issue_info.get("state", "closed")
            closed_at = issue_info.get("closedAt", "N/A")
        except Exception:
            title = "Unknown"
            state = "closed"
            closed_at = "N/A"

        # Find PRs linked to this sub-issue
        linked_prs: list[str] = []
        try:
            pr_result = subprocess.run(
                [
                    "gh", "pr", "list",
                    "--repo", repo,
                    "--state", "merged",
                    "--search", f"#{num} in:body",
                    "--json", "number,title,url",
                ],
                capture_output=True, text=True, timeout=15,
            )
            if pr_result.returncode == 0:
                prs = json.loads(pr_result.stdout) or []
                for pr in prs:
                    linked_prs.append(f"- PR #{pr['number']}: {pr['title']} ({pr['url']})")
        except Exception:
            pass

        # Also check open PRs
        try:
            pr_result2 = subprocess.run(
                [
                    "gh", "pr", "list",
                    "--repo", repo,
                    "--state", "open",
                    "--search", f"#{num} in:body",
                    "--json", "number,title,url",
                ],
                capture_output=True, text=True, timeout=15,
            )
            if pr_result2.returncode == 0:
                prs2 = json.loads(pr_result2.stdout) or []
                for pr in prs2:
                    linked_prs.append(f"- PR #{pr['number']}: {pr['title']} ({pr['url']})")
        except Exception:
            pass

        lines.append(f"### #{num}: {title}")
        lines.append(f"Status: {state}")

        if linked_prs:
            lines.append("Linked PRs:")
            lines.extend(linked_prs)
        else:
            lines.append("No linked PRs found.")

        lines.append("")

    lines.append("---")
    lines.append("All sub-issues are complete. Please review the results and aggregate them into a final summary.")
    return "\n".join(lines)


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Inject aggregated sub-issue results into a session JSON",
    )
    parser.add_argument("--session", required=True, help="Path to session.json")
    parser.add_argument("--repo", required=True, help="GitHub repo (OWNER/REPO)")
    parser.add_argument("--parent", required=True, type=int, help="Parent issue number")
    parser.add_argument("--sub-issues", required=True, help="Comma-separated sub-issue numbers")
    parser.add_argument("--out", required=True, help="Output path for modified session.json")
    args = parser.parse_args()

    # Parse sub-issue numbers
    sub_issues = [int(x.strip()) for x in args.sub_issues.split(",") if x.strip()]

    # Load session
    with open(args.session, "r") as f:
        session = json.load(f)

    messages: list[dict[str, Any]] = session.get("messages", [])
    last_tool_idx = find_last_tool_index(messages)

    if last_tool_idx is None:
        print("No tool message found in session. Appending as user message.", file=sys.stderr)
        # Fallback: append as user message
        summary = gather_sub_results(args.repo, sub_issues)
        messages.append({
            "role": "user",
            "content": summary,
        })
    else:
        # Replace the last tool message content
        summary = gather_sub_results(args.repo, sub_issues)
        messages[last_tool_idx]["content"] = summary

    session["messages"] = messages

    # Write output
    with open(args.out, "w") as f:
        json.dump(session, f, indent=2)

    print(f"Session updated: replaced tool message at index {last_tool_idx}", file=sys.stderr)
    print(f"Output written to {args.out}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: src/lib/inject-sub-results.ts
/**
 * inject-sub-results.ts — Replace the last tool message in a session with
 * aggregated sub-issue completion results, so the orchestrator sees a
 * summary of what happened when it's re-invoked for final aggregation.
 * The one canonical implementation (was inject_sub_results.ts), called
 * directly by aggregate_sub_issues.ts -- no more subprocess spawn.
 */
import { readFileSync, writeFileSync } from "node:fs";
import { gh } from "./gh.ts";
import type { GhPrSummary } from "./types.ts";
import type { Session, SessionMessage } from "./session.ts";

export type { Session, SessionMessage };

function findLastToolIndex(messages: SessionMessage[]): number | null {
  for (let i = messages.length - 1; i >= 0; i--) {
    if (messages[i]?.role === "tool") return i;
  }
  return null;
}

function gatherSubResults(repo: string, subIssues: number[]): string {
  const lines: string[] = ["All sub-issues have been completed.", "", "## Sub-issue Results", ""];

  for (const num of subIssues) {
    let title = "Unknown";
    // Not "closed". This block feeds the orchestrator's final report, whose whole
    // job is to state what happened, and a sub-issue whose state could not be
    // read is exactly the one thing that report must not assert. One rate-limited
    // lookup used to turn into "Status: closed" for work still in progress.
    let state = "could not be read";

    try {
      const { code, stdout } = gh("issue", "view", String(num), "--repo", repo, "--json", "title,state,closedAt");
      if (code === 0 && stdout) {
        const info = JSON.parse(stdout) as { title?: string; state?: string };
        title = info.title ?? "Unknown";
        state = info.state ?? "could not be read";
      }
    } catch {
      // keep defaults
    }

    const linkedPrs: string[] = [];
    // Tracked separately from an empty result, because "none found" and "could
    // not look" read identically once they reach the report and only one of them
    // is a fact.
    let prLookupFailed = false;
    for (const state_ of ["merged", "open"] as const) {
      try {
        const { code, stdout } = gh("pr", "list", "--repo", repo, "--state", state_, "--search", `#${num} in:body`, "--json", "number,title,url");
        if (code === 0 && stdout) {
          const prs = JSON.parse(stdout) as GhPrSummary[];
          for (const pr of prs) {
            linkedPrs.push(`- PR #${pr.number}: ${pr.title} (${pr.url})`);
          }
        } else {
          prLookupFailed = true;
        }
      } catch {
        prLookupFailed = true;
      }
    }

    lines.push(`### #${num}: ${title}`);
    lines.push(`Status: ${state}`);
    if (linkedPrs.length) {
      lines.push("Linked PRs:");
      lines.push(...linkedPrs);
    } else if (prLookupFailed) {
      lines.push("Linked PRs could not be read.");
    } else {
      lines.push("No linked PRs found.");
    }
    lines.push("");
  }

  lines.push("---");
  lines.push("All sub-issues are complete. Please review the results and aggregate them into a final summary.");
  return lines.join("\n");
}

/** Pure transform: returns an updated copy-in-place `session` with the last tool message replaced (or a new user message appended if none exists). */
export function injectSubResults(session: Session, repo: string, subIssues: number[]): Session {
  const messages = session.messages ?? [];
  const lastToolIdx = findLastToolIndex(messages);
  const summary = gatherSubResults(repo, subIssues);

  if (lastToolIdx === null) {
    console.error("No tool message found in session. Appending as user message.");
    messages.push({ role: "user", content: summary });
  } else {
    messages[lastToolIdx]!.content = summary;
  }

  session.messages = messages;
  return session;
}

/** File-based convenience wrapper matching the original CLI script's contract. */
export function injectSubResultsFile(sessionPath: string, repo: string, subIssues: number[], outPath: string): void {
  const session = JSON.parse(readFileSync(sessionPath, "utf8")) as Session;
  const updated = injectSubResults(session, repo, subIssues);
  writeFileSync(outPath, JSON.stringify(updated, null, 2));
}
